File: com-mini-util/src/main/java/com/mini/util/dao/IDao.java
package com.mini.util.dao;

import com.mini.util.dao.sql.SQLBuilder;
import com.mini.util.lang.ArraysUtil;
import com.mini.util.lang.Function;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

import static java.sql.ResultSet.CONCUR_READ_ONLY;
import static java.sql.ResultSet.TYPE_FORWARD_ONLY;
import static java.sql.Statement.RETURN_GENERATED_KEYS;

public interface IDao extends Connection {
    /**
     * 预处理对象填充
     * @param statement 预处理对象
     * @param params    参数
     * @throws SQLException 执行错误
     */
    default void fullPreparedStatement(PreparedStatement statement, Object[] params) throws SQLException {
        for (int i = 0; params != null && i < params.length; i++) {
            statement.setObject((i + 1), params[i]);
        }
    }

    /**
     * 批量操作
     * @param sql    SQL
     * @param length 数据长度
     * @param batch  回调接口
     * @return 执行结果
     * @throws SQLException 执行错误
     */
    default int[] batch(String sql, int length, Function.FR1<Object[], Integer> batch) throws SQLException {
        int[] result = new int[0];
        try (PreparedStatement statement = this.prepareStatement(sql)) {
            for (int index = 0; index < length; index++) {
                fullPreparedStatement(statement, batch.apply(index));
                // 添加 batch 数据
                statement.addBatch();

                // 提交 batch 数据
                if (index % 500 != 499) continue;
                result = ArraysUtil.concat(result, statement.executeBatch());
            }
            return ArraysUtil.concat(result, statement.executeBatch());
        }
    }

    /**
     * 批量操作
     * @param sql    SQL
     * @param params 参数
     * @return SQL执行结果
     * @throws SQLException 执行错误
     */
    default int[] batch(String sql, Object[][] params) throws SQLException {
        return batch(sql, params.length, index -> params[index]);
    }

    /**
     * 批量操作
     * @param sql        SQL
     * @param dataLength 数据长度
     * @param batch      回调接口
     * @return 执行结果
     * @throws SQLException 执行错误
     */
    default int[] batch(SQLBuilder sql, int dataLength, Function.FR1<Object[], Integer> batch) throws SQLException {
        return batch(sql.content(), dataLength, batch);
    }

    /**
     * 批量操作
     * @param sql    SQL
     * @param params 参数
     * @return 执行结果
     * @throws SQLException 执行错误
     */
    default int[] batch(SQLBuilder sql, Object[][] params) throws SQLException {
        return batch(sql.content(), params);
    }

    /**
     * 执行SQL
     * @param sql    SQL
     * @param params 参数
     * @return 执行结果
     * @throws SQLException 执行错误
     */
    default int execute(String sql, Object... params) throws SQLException {
        try (PreparedStatement statement = this.prepareStatement(sql)) {
            this.fullPreparedStatement(statement, params);
            return statement.executeUpdate();
        }
    }

    /**
     * 执行SQL
     * @param sql    SQL
     * @param params 参数
     * @return ID
     * @throws SQLException 执行错误
     */
    default long automatic(String sql, Object... params) throws SQLException {
        try (PreparedStatement statement = this.prepareStatement(sql, RETURN_GENERATED_KEYS)) {
            this.fullPreparedStatement(statement, params);
            statement.executeUpdate();

            // 读取自动生成的ID
            try (ResultSet rs = statement.getGeneratedKeys()) {
                return rs.next() ? rs.getLong(1) : 0;
            }
        }
    }

    /**
     * 执行SQL
     * @param sql SQL
     * @return 执行结果
     * @throws SQLException 执行错误
     */
    default int execute(SQL sql) throws SQLException {
        return execute(sql.content(), sql.params());
    }

    /**
     * 执行SQL
     * @param sql SQL
     * @return ID
     * @throws SQLException 执行错误
     */
    default long automatic(SQL sql) throws SQLException {
        return automatic(sql.content(), sql.params());
    }

    /**
     * 查询结果
     * @param sql    SQL
     * @param mapper 映射器
     * @param params 参数
     * @param <T>    解析器类型
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default <T> List<T> query(String sql, IMapper<T> mapper, Object... params) throws SQLException {
        try (PreparedStatement statement = prepareStatement(sql, TYPE_FORWARD_ONLY, CONCUR_READ_ONLY)) {
            IDao.this.fullPreparedStatement(statement, params);
            try (ResultSet resultSet = statement.executeQuery()) {
                List<T> result = new ArrayList<>();
                while (resultSet != null && resultSet.next()) {
                    int number = resultSet.getRow();
                    T t = mapper.execute(resultSet, number);
                    result.add(t);
                }
                return result;
            }
        }
    }

    /**
     * 查询结果
     * @param sql    SQL
     * @param mapper 映射器
     * @param <T>    解析器类型
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default <T> List<T> query(SQL sql, IMapper<T> mapper) throws SQLException {
        return query(sql.content(), mapper, sql.params());
    }

    /**
     * 查询结果
     * @param sql    SQL
     * @param mapper 映射器
     * @param params 参数
     * @param <T>    解析器类型
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default <T> T queryOne(String sql, IMapper<T> mapper, Object... params) throws SQLException {
        try (PreparedStatement statement = prepareStatement(sql, TYPE_FORWARD_ONLY, CONCUR_READ_ONLY)) {
            IDao.this.fullPreparedStatement(statement, params);
            try (ResultSet resultSet = statement.executeQuery()) {
                if (resultSet != null && resultSet.next()) {
                    int number = resultSet.getRow();
                    return mapper.execute(resultSet, number);
                }
                return null;
            }
        }
    }

    /**
     * 查询结果
     * @param sql    SQL
     * @param mapper 映射器
     * @param <T>    解析器类型
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default <T> T queryOne(SQL sql, IMapper<T> mapper) throws SQLException {
        return queryOne(sql.content(), mapper, sql.params());
    }

    /**
     * 查询String
     * @param sql    SQL
     * @param params 参数
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default String queryString(String sql, Object... params) throws SQLException {
        return queryOne(sql, (rs, n) -> rs.getString(1), params);
    }

    /**
     * 查询String
     * @param sql SQL
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default String queryString(SQL sql) throws SQLException {
        return queryString(sql.content(), sql.params());
    }

    /**
     * 查询Long
     * @param sql    SQL
     * @param params 参数
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default long queryLong(String sql, Object... params) throws SQLException {
        return queryOne(sql, (rs, n) -> rs.getLong(1), params);
    }

    /**
     * 查询Long
     * @param sql SQL
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default long queryLong(SQL sql) throws SQLException {
        return queryLong(sql.content(), sql.params());
    }

    /**
     * 查询Integer
     * @param sql    SQL
     * @param params 参数
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default int queryInt(String sql, Object... params) throws SQLException {
        return queryOne(sql, (rs, n) -> rs.getInt(1), params);
    }


    /**
     * 查询Integer
     * @param sql SQL
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default int queryInt(SQL sql) throws SQLException {
        return queryInt(sql.content(), sql.params());
    }


    /**
     * 查询Short
     * @param sql    SQL
     * @param params 参数
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default short queryShort(String sql, Object... params) throws SQLException {
        return queryOne(sql, (rs, n) -> rs.getShort(1), params);
    }

    /**
     * 查询Short
     * @param sql SQL
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default short queryShort(SQL sql) throws SQLException {
        return queryShort(sql.content(), sql.params());
    }


    /**
     * 查询Byte
     * @param sql    SQL
     * @param params 参数
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default byte queryByte(String sql, Object... params) throws SQLException {
        return queryOne(sql, (rs, n) -> rs.getByte(1), params);
    }

    /**
     * 查询Byte
     * @param sql SQL
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default byte queryByte(SQL sql) throws SQLException {
        return queryByte(sql.content(), sql.params());
    }

    /**
     * 查询Double
     * @param sql    SQL
     * @param params 参数
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default double queryDouble(String sql, Object... params) throws SQLException {
        return queryOne(sql, (rs, n) -> rs.getDouble(1), params);
    }

    /**
     * 查询Double
     * @param sql SQL
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default double queryDouble(SQL sql) throws SQLException {
        return queryDouble(sql.content(), sql.params());
    }

    /**
     * 查询Float
     * @param sql    SQL
     * @param params 参数
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default float queryFloat(String sql, Object... params) throws SQLException {
        return queryOne(sql, (rs, n) -> rs.getFloat(1), params);
    }

    /**
     * 查询Float
     * @param sql SQL
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default float queryFloat(SQL sql) throws SQLException {
        return queryFloat(sql.content(), sql.params());
    }

    /**
     * 查询Boolean
     * @param sql    SQL
     * @param params 参数
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default boolean queryBoolean(String sql, Object... params) throws SQLException {
        return queryOne(sql, (rs, n) -> rs.getBoolean(1), params);
    }

    /**
     * 查询Boolean
     * @param sql SQL
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default boolean queryBoolean(SQL sql) throws SQLException {
        return queryBoolean(sql.content(), sql.params());
    }

    /**
     * 查询Timestamp
     * @param sql    SQL
     * @param params 参数
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default Timestamp queryTimestamp(String sql, Object... params) throws SQLException {
        return queryOne(sql, (rs, n) -> rs.getTimestamp(1), params);
    }

    /**
     * 查询Timestamp
     * @param sql SQL
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default Timestamp queryTimestamp(SQL sql) throws SQLException {
        return queryTimestamp(sql.content(), sql.params());
    }

    /**
     * 查询Date
     * @param sql    SQL
     * @param params 参数
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default Date queryDate(String sql, Object... params) throws SQLException {
        return queryOne(sql, (rs, n) -> rs.getDate(1), params);
    }

    /**
     * 查询Date
     * @param sql SQL
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default Date queryDate(SQL sql) throws SQLException {
        return queryDate(sql.content(), sql.params());
    }

    /**
     * 查询Time
     * @param sql    SQL
     * @param params 参数
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default Time queryTime(String sql, Object... params) throws SQLException {
        return queryOne(sql, (rs, n) -> rs.getTime(1), params);
    }

    /**
     * 查询Time
     * @param sql SQL
     * @return 查询结果
     * @throws SQLException 执行错误
     */
    default Time queryTime(SQL sql) throws SQLException {
        return queryTime(sql.content(), sql.params());
    }


    /**
     * 查询结果
     * @param paging paging 分页器
     * @param sql    SQL
     * @param row    解析器
     * @param params 参数
     * @param <T>    解析器类型
     * @return 查询结果
     */
    default <T> List<T> query(Paging paging, String sql, IMapper<T> row, Object... params) throws SQLException {
        paging.setTotal(this.queryInt(this.totals(sql), params));
        return query(paging(paging, sql), row, params);
    }

    /**
     * 查询结果
     * @param paging 分页器
     * @param sql    SQL
     * @param row    解析器
     * @param <T>    解析器类型
     * @return 查询结果
     */
    default <T> List<T> query(Paging paging, SQL sql, IMapper<T> row) throws SQLException {
        return query(paging, sql.content(), row, sql.params());
    }

    /**
     * 事件执行方法
     * @param transaction 事务回调
     * @return true-提交； false-回滚
     */
    default boolean transaction(Function.FR1<Boolean, IDao> transaction) throws SQLException {
        return transaction(TRANSACTION_REPEATABLE_READ, transaction);
    }

    /**
     * 事件执行方法
     * @param level       事务级别
     * @param transaction 事务回调
     * @return true-提交； false-回滚
     */
    boolean transaction(int level, Function.FR1<Boolean, IDao> transaction) throws SQLException;

    /**
     * 返回查询分页总条数的SQL
     * @param sql SQL
     * @return 查询总条数的SQL
     */
    String totals(String sql);

    /**
     * 根据分页参数，组装分页查询SQL
     * @param paging 分页参数
     * @param sql    基础查询SQL
     * @return 分页查询SQL
     */
    String paging(Paging paging, String sql);
}

File: sn.mini.kotlin.jdbc/src/main/java/sn/mini/kotlin/jdbc/IDao.kt
package sn.mini.kotlin.jdbc

import sn.mini.kotlin.jdbc.sql.Sql
import sn.mini.kotlin.util.Paging
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.util.*

interface IDao : java.sql.Connection {
    /**
     * 填充预处理sql语句
     * @param statement 预处理sql 对象
     * @param params statement 对象中 sql对应的参数值
     */
    fun fullPreparedStatement(statement: PreparedStatement, vararg params: Any)

    /**
     * 批量SQL操作实现方法<br></br>
     * @param sql 预编译SQL语句
     * @param length 数据长度
     * @param values 回调接口
     */
    fun batch(sql: String, length: Int, values: ((index: Int) -> Array<Any>)): IntArray

    /**
     * 批量SQL操作实现方法
     * @param sql
     * @param params
     * @return
     */
    fun batch(sql: String, params: Array<Array<Any>>): IntArray

    /**
     * 批量SQL操作实现方法<br></br>
     * @param sql 预编译SQL语句
     * @param length 数据长度
     * @param values 回调接口
     */
    fun batch(sql: Sql, length: Int, values: (index: Int) -> Array<Any>): IntArray

    /**
     * 批量SQL操作实现方法
     * @param sql
     * @param params
     * @return
     */
    fun batch(sql: Sql, params: Array<Array<Any>>): IntArray

    /**
     * 执行insert,replace,update,delete语句,返回影响数据库记录的行数
     * @param sql
     * @param params
     * @return
     */
    fun execute(sql: String, vararg params: Any): Int

    /**
     * 执行insert,replace语句, 返回数据库自增ID
     * @param sql
     * @param params
     * @return 返回自增长ID值
     */
    fun executeGen(sql: String, vararg params: Any): Long

    /**
     * 执行insert,replace,update,delete语句,返回影响数据库记录的行数
     * @param sql
     * @return
     */
    fun execute(sql: Sql): Int

    /**
     * 执行insert,replace语句, 返回数据库自增ID
     * @param sql
     * @return 返回自增长ID值
     */
    fun executeGen(sql: Sql): Long

    /**
     * 生成SQL结果: insert into user_info(coumn1, column2, column3) values(?, ?, ?)
     * @param tableName 表名称 比如: user_info
     * @param keys 比如: coumn1, column2, column3
     * @param params 比如: va1, va2, va3 分别对应三个字段的值
     * @return
     */
    fun insert(tableName: String, keys: Array<String>, vararg params: Any): Int

    /**
     * 生成SQL结果: insert into user_info(coumn1, column2, column3) values(?, ?, ?)
     * @param tableName 表名称 比如: user_info
     * @param keys 比如: coumn1, column2, column3
     * @param params 比如: va1, va2, va3 分别对应三个字段的值
     * @return 返回自增长ID值
     */
    fun insertGen(tableName: String, keys: Array<String>, vararg params: Any): Long

    /**
     * 生成SQL结果: replace into user_info(coumn1, column2, column3) values(?, ?, ?)
     * @param tableName 表名称 比如: user_info
     * @param keys 比如: coumn1, column2, column3
     * @param params 比如: va1, va2, va3 分别对应三个字段的值
     * @return
     */
    fun replace(tableName: String, keys: Array<String>, vararg params: Any): Int

    /**
     * 生成SQL结果: replace into user_info(coumn1, column2, column3) values(?, ?, ?)
     * @param tableName 表名称 比如: user_info
     * @param keys 比如: coumn1, column2, column3
     * @param params 比如: va1, va2, va3 分别对应三个字段的值
     * @return 返回自增长ID值
     */
    fun replaceGen(tableName: String, keys: Array<String>, vararg params: Any): Long

    /**
     * 生成SQL结果: update coumn1 = ?, column2 = ?, column3 = ? where coumn4 = ? and column5 = ?
     * @param tableName 表名称 比如: user_info
     * @param keys 比如: coumn1, column2, column3
     * @param wheres 比如: coumn4, column5 默认用and方式连接
     * @param params 比如: va1, va2, va3, va4, va5, 其中前三个对应keys的值, 后两个对应wheres的值
     * @return
     */
    fun update(tableName: String, keys: Array<String>, wheres: Array<String>, vararg params: Any): Int

    /**
     * 生成SQL结果: delete user_info where coumn1 = ? and column2 = ? and column3 = ?
     * @param tableName 表名称 比如: user_info
     * @param wheres 比如: column1, column2, column3 默认用and方式连接
     * @param params 比如: va1, va2, va3 分别对应三个字段值
     * @return 影响数据库的条数
     */
    fun delete(tableName: String, wheres: Array<String>, vararg params: Any): Int

    /**
     * 生成SQL结果: delete user_info where coumn1 = ? and column2 = ? and column3 = ?
     * @param tableName 表名称 比如: user_info
     * @param wheres 条件字段，一般用于ID字段
     * @param params 比如: va1, va2, va3 分别对应三个字段值
     * @return 影响数据库的条数
     */
    fun delete(tableName: String, wheres: String, vararg params: Any): Int

    /**
     * 查询记录, 根据sql中所有带条查询数据库所有匹配的记录<br></br>
     * 调用该方法时, 一定要手动关闭ResultSet结果集,否则导致无法关闭连接
     * @param row 回调解析过程
     * @param sql sql语句
     * @param params sql语句是的参数
     * @return
     */
    fun <T> executeQuery(row: (rs: ResultSet) -> T?, sql: String, vararg params: Any): T

    /**
     * 查询记录, 根据sql所所带条件查询数据库所有匹配的记录<br></br>
     * 调用该方法时, 一定要手动关闭ResultSet结果集,否则导致无法关闭连接
     * @param row 回调解析过程
     * @param sql sql封装器
     * @return
     */
    fun <T> executeQuery(row: (rs: ResultSet) -> T?, sql: Sql): T?

    /**
     * 查询记录, 根据sql所所带条件查询数据库记录按分页工具分页<br></br>
     * 调用该方法时, 一定要手动关闭ResultSet结果集,否则导致无法关闭连接
     * @param row 回调解析过程
     * @param paging 分页工具
     * @param sql sql语句
     * @param params sql语句是的参数
     * @return
     */
    fun <T> executeQuery(row: (rs: ResultSet) -> T?, paging: Paging, sql: String, vararg params: Any): T?

    /**
     * 查询记录, 根据sql所所带条件查询数据库记录按分页工具分页<br></br>
     * 调用该方法时, 一定要手动关闭ResultSet结果集,否则导致无法关闭连接
     * @param row 回调解析过程
     * @param paging 分页工具
     * @param sql sql封装器
     * @return
     */
    fun <T> executeQuery(row: (rs: ResultSet) -> T?, paging: Paging, sql: Sql): T?

    /**
     * 查询记录， 根据sql中所带的条件查询数据库所有匹配的记录全部返回
     * @param sql sql封装工具
     * @return
     */
    fun query(sql: Sql): ArrayList<HashMap<String, Any>?>

    /**
     * 根据 查询所有记录
     * @param sql
     * @param params
     * @return
     */
    fun query(sql: String, vararg params: Any): ArrayList<HashMap<String, Any>?>

    /**
     * 查询记录， 根据sql中所带的条件查询数据库所有匹配的记录全部返回
     * @param clazz 实体对象class 为了节约性能 ，实体对象属性中不能有复合对象，只有能简单数据 类型，否则组装失败
     * @param sql 查询记录， 根据sql中所带的条件查询数据库所有匹配的记录全部返回
     * @return 查询结果
     */
    fun <T> query(clazz: Class<T>, sql: Sql): List<T?>

    /**
     * 查询所有记录
     * @param clazz 实体对象class. 为了节约性能 ，实体对象属性中不能有复合对象，只有能简单数据 类型，否则组装失败
     * @param sql 查询sql语句
     * @param params
     * @return
     */
    fun <T> query(clazz: Class<T>, sql: String, vararg params: Any): List<T?>

    /**
     * 查询记录， 根据sql中所带的条件查询数据库所有匹配的记录全部返回, 该方法中sql中的cellMapper无效
     * @param mapper
     * @param sql
     * @return
     */
    fun <T> query(mapper: (rs: ResultSet) -> T?, sql: Sql): List<T?>

    /**
     * 查询所有数据
     * @param mapper
     * @param sql
     * @param params
     * @return
     */
    fun <T> query(mapper: (rs: ResultSet) -> T?, sql: String, vararg params: Any): List<T?>

    /**
     * 查询记录， 根据sql中所带的条件与Paging工具分页查询数据库
     * @param paging 分页工具
     * @param sql sql封装工具
     * @return 查询结果 @see com.ms.mvc.util.JSONArrays
     */
    fun query(paging: Paging, sql: Sql): ArrayList<HashMap<String, Any>?>

    /**
     * 分页查询数据
     * @param paging
     * @param sql
     * @param params
     * @return
     */
    fun query(paging: Paging, sql: String, vararg params: Any): ArrayList<HashMap<String, Any>?>

    /**
     * 查询记录， 根据sql中所带的条件与Paging工具分页查询数据库
     * @param paging sql 分页工具
     * @param clazz 实体对象class. 为了节约性能 ，实体对象属性中不能有复合对象，只有能简单数据 类型，否则组装失败
     * @param sql
     * @return
     */
    fun <T> query(paging: Paging, clazz: Class<T>, sql: Sql): List<T?>

    /**
     * 查询所有记录
     * @param paging
     * @param clazz 实体对象class. 为了节约性能 ，实体对象属性中不能有复合对象，只有能简单数据 类型，否则组装失败
     * @param sql
     * @param params
     * @return
     */
    fun <T> query(paging: Paging, clazz: Class<T>, sql: String, vararg params: Any): List<T?>

    /**
     * 查询记录， 根据sql中所带的条件查询数据库所有匹配的记录全部返回, 该方法中sql中的cellMapper无效
     * @param paging
     * @param mapper
     * @param sql
     * @return
     */
    fun <T> query(paging: Paging, mapper: (rs: ResultSet) -> T?, sql: Sql): List<T?>

    /**
     * 分页查询数据
     * @param paging
     * @param mapper
     * @param sql
     * @param params
     * @return
     */
    fun <T> query(paging: Paging, mapper: (rs: ResultSet) -> T?, sql: String, vararg params: Any): List<T?>

    /**
     * 查询单条记录
     * @param sql sql封装工具
     * @return 查询结果 @see com.ms.mvc.util.JSONArrays
     */
    fun queryOne(sql: Sql): HashMap<String, Any>?

    /**
     * 查询单条记录
     * @param sql
     * @param params
     * @return
     */
    fun queryOne(sql: String, vararg params: Any): HashMap<String, Any>?

    /**
     * 查询单条记录
     * @param  clazz 为了节约性能 ，实体对象属性中不能有复合对象，只有能简单数据 类型，否则组装失败
     * @param sql  根据sql中所带的条件查询数据库所有匹配的记录全部返回
     * @return 查询结果
     */
    fun <T> queryOne(clazz: Class<T>, sql: Sql): T?

    /**
     * 查询单条记录
     * @param clazz 实体对象class. 为了节约性能 ，实体对象属性中不能有复合对象，只有能简单数据 类型，否则组装失败
     * @param sql
     * @param params
     * @return
     */
    fun <T> queryOne(clazz: Class<T>, sql: String, vararg params: Any): T?

    /**
     * 查询单条记录 该方法中sql中的cellMapper无效
     * @param mapper
     * @param sql
     * @return
     */
    fun <T> queryOne(mapper: (rs: ResultSet) -> T?, sql: Sql): T?

    /**
     * 查询单条记录
     * @param mapper
     * @param sql
     * @param params
     * @return
     */
    fun <T> queryOne(mapper: (rs: ResultSet) -> T?, sql: String, vararg params: Any): T?

    /**
     * 查询单一 String 数据值
     * @param sql
     */
    fun queryString(sql: Sql): String?

    /**
     * 查询单一 String 数据值
     * @param sql
     * @param params
     * @return
     */
    fun queryString(sql: String, vararg params: Any): String?

    /**
     * 查询单一 Long 数据值
     * @param sql
     */
    fun queryLong(sql: Sql): Long

    /**
     * 查询单一 Long 数据值
     * @param sql
     * @param params
     * @return
     */
    fun queryLong(sql: String, vararg params: Any): Long

    /**
     * 查询单一 int 数据值
     * @param sql
     */
    fun queryInt(sql: Sql): Int

    /**
     * 查询单一 int 数据值
     * @param sql
     * @param params
     * @return
     */
    fun queryInt(sql: String, vararg params: Any): Int

    /**
     * 查询单一 short 数据值
     * @param sql
     */
    fun queryShort(sql: Sql): Short

    /**
     * 查询单一 short 数据值
     * @param sql
     * @param params
     * @return
     */
    fun queryShort(sql: String, vararg params: Any): Short

    /**
     * 查询单一 byte 数据值
     * @param sql
     */
    fun queryByte(sql: Sql): Byte

    /**
     * 查询单一 short 数据值
     * @param sql
     * @param params
     * @return
     */
    fun queryByte(sql: String, vararg params: Any): Byte

    /**
     * 查询单一 double 数据值
     * @param sql
     */
    fun queryDouble(sql: Sql): Double

    /**
     * 查询单一 double 数据值
     * @param sql
     * @param params
     * @return
     */
    fun queryDouble(sql: String, vararg params: Any): Double

    /**
     * 查询单一 float 数据值
     * @param sql
     */
    fun queryFloat(sql: Sql): Float

    /**
     * 查询单一 float 数据值
     * @param sql
     * @param params
     * @return
     */
    fun queryFloat(sql: String, vararg params: Any): Float

    /**
     * 查询单一 date 数据值
     * @param sql
     */
    fun queryDate(sql: Sql): Date

    /**
     * 查询单一 date 数据值
     * @param sql
     * @param params
     * @return
     */
    fun queryDate(sql: String, vararg params: Any): Date?

    /**
     * 查询单一 boolean 数据值
     * @param sql
     */
    fun queryBoolean(sql: Sql): Boolean

    /**
     * 查询单一 boolean 数据值
     * @param sql
     * @param params
     * @return
     */
    fun queryBoolean(sql: String, vararg params: Any): Boolean

    /**
     * 事件执行方法
     * @param transaction
     * @return true: 事务执行成功(最后提交事务), false: 事务执行失败, 最后回滚
     */
    fun transaction(transaction: (() -> Boolean)): Boolean

    /**
     * 事件执行方法
     * @param transLevel 事务级别
     * @param transaction
     * @return true: 事务执行成功(最后提交事务), false: 事务执行失败, 最后回滚
     */
    fun transaction(transLevel: Int, transaction: (() -> Boolean)): Boolean

}